### DISCUSSION TITLE
Hiding a category on Model tree does not hide them in categories tree

### DISCUSSION THREAD

**Discussion by muthubentley** [iTwin/itwinjs-core#5151] [Q&A] [answered]
I hide three categories "3DWINDOW,3DWSWING,ANNOBJ" in Models tree
![image](https://user-images.githubusercontent.com/107229931/222032255-37c424aa-ef0f-462a-a0eb-80729d61fcdb.png)

Though the categories are hidden in the viewer, when I switch to Categories tree, these categories are in visible state
![image](https://user-images.githubusercontent.com/107229931/222032760-99a20232-a9bf-4651-8094-d7e757937cc5.png)

viewport.view.categorySelector.categories also include the hidden categories.

This can be reproduced in the below sandbox:
https://www.itwinjs.org/sandboxes/MuthuKalamani/ModelTreeWidgetCustomizations

**Comment by grigasp** [comment, marked as answer]
I believe that works as designed, although I agree it's very unintuitive. The Categories tree uses categories selector to control visibility and the Models Tree uses per-category-model-overrides. That's because the Models tree shows categories under models and we don't want to change category visibility for all models - we just want to turn it on/off for specific model it's under.

**Reply by muthubentley** [reply]
@grigasp  Thanks for the explanation. We found this problem while creating a saved view. When I hide a category in Models tree and create a saved view, I pass "viewport.view.categorySelector.categories" to the Savedviews api. Since the category I hide is still present in  "viewport.view.categorySelector.categories", loading a saved view does not give the exact same view as before. Is there any other way to get only the visible categories ?

**Reply by grigasp** [reply]
@pmconne, is it possible to persist per-model-category overrides' information into a saved view?

**Reply by pmconne** [reply]
@muthubentley please be specific about what "Saved Views API" you are using.

**Reply by muthubentley** [reply]
@grigasp  @pmconne 
Fukui is using 
https://developer.bentley.com/apis/savedviews/operations/create-savedview/
In request payload , below array of enabled categories is set from "viewport.view.categorySelector.categories". 
"itwin3dView": {
 "categories": {
                "enabled": [
                    "0xxx",
                    "0xxxx"
                ],               
            },
}

The problem is even I hide a category in the Models tree, that category is not excluded from  "viewport.view.categorySelector.categories". Hence When the saved view is loaded, that category does not get hidden in the viewer.

**Reply by pmconne** [reply]
This is a question for the maintainers of the saved view service. @diegopinate @Laisvune please review. Also, does a TypeScript API (or plans to create one) exist to simplify saved view creation via this service by automatically serializing/deserializing the payload?

Other applications like Design Review also expect the per-model category visibility to be preserved when creating and recalling a saved view. @jjbeckman13 @jason-crow as you transition to the saved views service have you observed the same problem? Are you working around it by, e.g., using an [extension](https://developer.bentley.com/apis/savedviews/operations/create-savedview/#extensionpayload)?

@nick4598 I seem to recall you made some potentially related changes in itwinjs-core for serializing per-model category visibility, but I can't recall the details - can you?

**Reply by muthubentley** [reply]
@diegopinate @Laisvune  awaiting your reply.

**Reply by jason-crow** [reply]
My understanding is that it will be preserved via an extension in the new service. @diegopinate can  provide details

**Reply by diegopinate** [reply]
Yes, extensions are the way to go here. Muthu has been emailing me and he has used extensions successfully now. In this case, "PerModelCategoryVisibility" must be used.

See [Create Extension](https://developer.bentley.com/apis/savedviews/operations/create-extension/)

**Reply by nick4598** [reply]
@pmconne Sorry for the late response, 

The main change which I made surrounding this was to provide a 'setOverrides' function in addition to the already present 'setOverride' function. setOverrides would just be a slightly more efficient way for frontends to communicate with the backend since currently calling setOverride for each individual PerModelCategoryVisibility would be one round trip to the backend to actually load the subcategories. setOverrides would be less round trips between frontend and backend.

**Reply by muthubentley** [reply]
Thanks all for the reply. Understood that this can be achieved using extensions. @nick4598 Thanks for 'setOverrides', it is very useful

**Comment by jjbeckman13** [comment]
@pmconne, we have not transitioned yet to the new service. A member of Diego's team is integrating a client for the new API into our existing Ui component. We have not had the chance to use it.

I would expect per-model category visibility to be preserved when recreating and recalling a saved view.
